torch_geometric/loader/link_neighbour_loader.py

```diff
@@ -0,0 +1,95 @@
+from typing import Any, Callable, List, Tuple
+
+import torch
+from torch import Tensor
+
+from torch_geometric.data import Data
+from torch_geometric.loader.base import BaseDataLoader
+from torch_geometric.loader.neighbor_loader import NeighborSampler
+from torch_geometric.loader.utils import filter_data
+from torch_geometric.typing import InputEdges, NumNeighbors, OptTensor
+from torch_geometric.utils.mask import mask_to_index
+
+
+class LinkNeighborSampler(NeighborSampler):
+    def __init__(
+        self,
+        data: Data,
+        num_neighbors: NumNeighbors,
+        replace: bool = False,
+        directed: bool = True,
+    ):
+        super().__init__(data, num_neighbors, replace, directed)
+        self.data = data
+
+    def __call__(self, index: List[Tensor]):
+
+        batch_size = len(index)
+
+        # take start and end node from each edge then deduplicate
+        row, col = self.data.edge_index
+        node_index = torch.cat([row[index], col[index]], dim=0)
+        node_index = torch.unique(node_index)
+
+        # get sampled graph
+        node, row, col, edge, _ = super().__call__(index)
+        return node, row, col, edge, batch_size, index
+
+
+class LinkNeighborLoader(BaseDataLoader):
+    def __init__(
+        self,
+        data: Data,
+        num_neighbors: NumNeighbors,
+        input_edge_idx: InputEdges = None,
```

Comment:
```suggestion
        input_edges: InputEdges = None,
```
Aligns better with `NeighborLoader`. WDYT?

Reply:
The reason I changed this was that I need to eventually recover the index of the edges to look up any edge attributes that exist. So if taking as input the actual edges I need to do a search for each edge in the `data.edge_index`.

But... it does seem somehow more usable if you provide edges, so I'm a fan of your suggestion with the caveat that we'll need to do a mapping from edges -> edges_index.

Updating the PR with this, but let me know if you had something else in mind.

Reply:
Actually, I can't see an efficient way to do this... leaving both in the MR.. but I'm hesitant to keep the format without 
it explicitly being an index.

Reply:
Going to resolve this and move it as this file is gone (renamed)